File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {
    int maxVisitNum = 0;
    
    private void dfs(boolean[] visited, int visitNum, int k, int[][] dungeons){
        maxVisitNum = Math.max(maxVisitNum, visitNum);
        
        for(int i = 0; i < visited.length; i++){
            if(visited[i]) continue; // 이미 방문한 곳
            if(k < dungeons[i][0]) continue; // 최소 피로도 조건에 부합X
            
            visited[i] = true;
            dfs(visited, visitNum + 1, k - dungeons[i][1], dungeons);
            visited[i] = false;
        }
    }
    
    public int solution(int k, int[][] dungeons) {
        boolean[] visited = new boolean[dungeons.length];
        dfs(visited, 0, k, dungeons);
        
        return maxVisitNum;
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    global maxCount
    maxCount = 0
    visited = [False for _ in range(len(dungeons))]
    
    def exploring(k, visited, dungeons, count):
        global maxCount
        maxCount = max(count, maxCount)
        if(len(visited) == dungeons):
            return

        for i in range(len(dungeons)):
            minFatigue, useFatigue = dungeons[i]
            if(visited[i]): continue
            if(k < minFatigue): continue

            visited[i] = True
            exploring(k-useFatigue, visited, dungeons, count+1)
            visited[i] = False
    
    exploring(k, visited, dungeons, 0)
    
    return maxCount